### PR TITLE
k8s: Add support for Kubernetes 1.30.0

### PR DESCRIPTION
**Pull request by christarazi**
Update to k8s 1.30.0 and fix code-generator scripts.

Also, bump sigs.k8s.io/controller-runtime to the latest main version as
of this commit (v0.17.1-0.20240327193027-21368602d84b) to workaround a
breaking change introduced in 1.30 libraries.

Co-authored-by: André Martins <andre@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

### CODE REVIEW

File: .github/actions/aws/k8s-versions.yaml
Comment (aanm) [inline .github/actions/aws/k8s-versions.yaml]: I don't think this will work because this version is not available yet.

File: .github/actions/ginkgo/main-focus.yaml
Comment (aanm) [inline .github/actions/ginkgo/main-focus.yaml:249]:
```suggestion
  # These tests require an external node which is only available on 1.29
```

File: .github/actions/ginkgo/main-k8s-versions.yaml
Comment (aanm) [inline .github/actions/ginkgo/main-k8s-versions.yaml]:
```suggestion
    kube-image: "quay.io/cilium/kindest-node:v1.30.0-rc.0@sha256:98fff7e882a1c93e9d7e06143476e982a2916009182c7f019eb95e5cda7cda0d"
```

File: .github/actions/ginkgo/main-k8s-versions.yaml
Comment (aanm) [inline .github/actions/ginkgo/main-k8s-versions.yaml]:
```suggestion
    kube-image: "kindest/node:v1.29.2@sha256:51a1434a5397193442f0be2a297b488b6c919ce8a3931be0ce822606ea5ca245"
```

File: .github/actions/ginkgo/main-k8s-versions.yaml
Comment (aanm) [inline .github/actions/ginkgo/main-k8s-versions.yaml]:
```suggestion
    kube-image: "kindest/node:v1.28.7@sha256:9bc6c451a289cf96ad0bbaf33d416901de6fd632415b076ab05f5fa7e4f65c58"
```

File: .github/actions/ginkgo/main-k8s-versions.yaml
Comment (aanm) [inline .github/actions/ginkgo/main-k8s-versions.yaml]:
```suggestion
  - k8s-version: "1.27"
    ip-family: "dual"
    # renovate: datasource=docker
    kube-image: "kindest/node:v1.27.11@sha256:681253009e68069b8e01aad36a1e0fa8cf18bb0ab3e5c4069b2e65cafdd70843"
    # renovate: datasource=docker depName=quay.io/lvh-images/kind
    kernel: "5.4-20240305.092417@sha256:054026744db7305801641f276d751d6179dc5eed9a1eaa4c8b6918982e5f828e"
```

File: Makefile
Comment (aanm) [inline Makefile]: Do we need these changes? I know that kube_codegen.sh was introduced but if the older script it's still available let's use it. Changeing to the `kube_codegen.sh` will require more changes IIRC.

File: test/provision/k8s_install.sh
Comment (aanm) [inline test/provision/k8s_install.sh:599]: This should be `1.30.0-rc.0`

**Review by aanm — CHANGES_REQUESTED** [review]

File: Makefile
Comment (sayboras) [reply Makefile]: The old script is available, but didn't do anything right ? :thinking: 

https://github.com/kubernetes/code-generator/blob/v0.30.0-rc.1/generate-internal-groups.sh

> echo "ERROR: $(basename "$0") has been removed."
echo "ERROR: Please use k8s.io/code-generator/kube_codegen.sh instead."

Comment (christarazi) [reply Makefile]: Yes it was removed and we had to change the scripts to use `kube_codegen.sh`.

File: pkg/k8s/slim/k8s/apis/meta/v1/types.go
Comment (aanm) [inline pkg/k8s/slim/k8s/apis/meta/v1/types.go]: Why did you remove the deepequals? :thinking:

File: pkg/k8s/slim/k8s/apis/meta/v1/types.go
Comment (aanm) [inline pkg/k8s/slim/k8s/apis/meta/v1/types.go:240]: I think we need to leave the `MatchLabelsValue` because of the schema validation, is that not necessary anymore?

File: pkg/k8s/slim/k8s/apis/meta/v1/types.go
Comment (aanm) [inline pkg/k8s/slim/k8s/apis/meta/v1/types.go]: I think we can remove this entire structure since it's not used anywhere.

**Review by aanm — CHANGES_REQUESTED** [review]
There are somethings I noticed on a first pass. I'll do a second pass tomorrow.

File: .github/actions/ginkgo/main-focus.yaml
Comment (aanm) [inline .github/actions/ginkgo/main-focus.yaml]: oh, this space is needed, that's how the test is named. Otherwise the linter will fail.

**Review by aanm — CHANGES_REQUESTED** [review]

File: .github/actions/ginkgo/main-focus.yaml
Comment (christarazi) [reply .github/actions/ginkgo/main-focus.yaml]: Oh good catch. My editor prunes extra whitespace on save so I didn't intend to do this.

File: pkg/k8s/slim/k8s/apis/meta/v1/types.go
Comment (christarazi) [reply pkg/k8s/slim/k8s/apis/meta/v1/types.go]: This was unintended. Re-adding.

File: pkg/k8s/slim/k8s/apis/meta/v1/types.go
Comment (christarazi) [reply pkg/k8s/slim/k8s/apis/meta/v1/types.go:240]: Ah, you're right. I also added that as an extra gotcha in the documention commit.

**Review by ghost — APPROVED** [review]
docs good

File: pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwideenvoyconfigs.yaml
Comment (aanm) [inline pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwideenvoyconfigs.yaml]: What happened here?

File: pkg/k8s/slim/k8s/apis/meta/v1/types.go
Comment (aanm) [inline pkg/k8s/slim/k8s/apis/meta/v1/types.go]: I think this was accidentally removed

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/k8s/slim/k8s/apis/meta/v1/types.go
Comment (christarazi) [reply pkg/k8s/slim/k8s/apis/meta/v1/types.go]: Yep, my mistake. Another example of something to keep in mind when updating the slim structures. There are lots of these little changes that we've made over the years that we depend on. Now I'm documenting them in the README of the slim directory.

File: pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwideenvoyconfigs.yaml
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwideenvoyconfigs.yaml]: Fixed

File: pkg/k8s/slim/k8s/apis/meta/v1/types.go
Comment (christarazi) [reply pkg/k8s/slim/k8s/apis/meta/v1/types.go]: See https://github.com/cilium/cilium/compare/af14f15fb803cc01c117206dc43ea057f6ad2ca9..4afcaa19c2852acdc40c125ed2817a15cd08ef49

**Review by aanm — APPROVED** [review]

**Comment by christarazi** [comment]
@tklauser @jschwinger233, could I get your sign off here?

File: Makefile
Comment (joestringer) [inline Makefile:350]: This macro now receives the `TMPDIR` as a parameter, but ignores it.

Why do we have the TMPDIR at all? I figure it's so that we can use the existing files in the tree to generate a bunch of new files in an entirely separate path, then iff all generation succeeds, we copy the full path back into the main directory. Seems a bit odd to be now generating files somewhere within the current directory instead?

File: Makefile
Comment (joestringer) [inline Makefile:400]: nit: Given that there are various steps inside this Make target which generate files within the TMPDIR, I don't think it makes sense to hide the copy inside the `k8s-code-gen.sh` script.

File: Makefile
Comment (joestringer) [reply Makefile:350]: Is the base problem here that `go-to-protobuf` removed the parameters we're relying on and refuses to recreate the directory structure in the output dir based on the input directory structure, thereby forcing us to directly inject the files back into the same tree?

**Review by joestringer — COMMENTED** [review]
Some minor nits and observations from the Make target pieces. I feel like the make targets we have here are fighting with the upstream k8s codegen tooling a bit. Non-blocking feedback, we can iterate further in-tree.

File: Makefile
Comment (joestringer) [reply Makefile:350]: Ugh yeah looks like upstream broke the ability to output somewhere outside `$GOPATH`. I see other commands adding a `--output-package` to provide the equivalent `$GOPATH` path presumably to infer something about the output, but they didn't bother for this command.

The best I'm coming up with so far as an alternative here is accept dirtying the tree and use this instead:
```
    --output-dir=$$GOPATH/src
```

I'll carry this change in https://github.com/cilium/cilium/pull/32063 since it'll be a dependency there if you don't integrate it in before merge.

**Comment by joestringer** [comment]
Are these errors printed in the (passing) `Generate k8s API` CI workflow a problem?

```
contrib/scripts/k8s-code-gen.sh "/tmp/cilium.tmphCFo2Q9B"
SCRIPT_ROOT=/home/runner/work/cilium/cilium/src/github.com/cilium/cilium/contrib/scripts/../..
Generating client code for 4 targets
grep: /tmp/cilium.tmphCFo2Q9B/github.com/cilium/cilium/pkg/k8s/slim/k8s/client/clientset: No such file or directory
Generating lister code for 4 targets
grep: /tmp/cilium.tmphCFo2Q9B/github.com/cilium/cilium/pkg/k8s/slim/k8s/client/listers: No such file or directory
Generating informer code for 4 targets
grep: /tmp/cilium.tmphCFo2Q9B/github.com/cilium/cilium/pkg/k8s/slim/k8s/client/informers: No such file or directory
Generating client code for 1 targets
grep: /tmp/cilium.tmphCFo2Q9B/github.com/cilium/cilium/pkg/k8s/slim/k8s/apiextensions-client/clientset: No such file or directory
Generating lister code for 1 targets
grep: /tmp/cilium.tmphCFo2Q9B/github.com/cilium/cilium/pkg/k8s/slim/k8s/apiextensions-client/listers: No such file or directory
Generating informer code for 1 targets
grep: /tmp/cilium.tmphCFo2Q9B/github.com/cilium/cilium/pkg/k8s/slim/k8s/apiextensions-client/informers: No such file or directory
Generating client code for 2 targets
grep: /tmp/cilium.tmphCFo2Q9B/github.com/cilium/cilium/pkg/k8s/client/clientset: No such file or directory
Generating lister code for 2 targets
grep: /tmp/cilium.tmphCFo2Q9B/github.com/cilium/cilium/pkg/k8s/client/listers: No such file or directory
Generating informer code for 2 targets
grep: /tmp/cilium.tmphCFo2Q9B/github.com/cilium/cilium/pkg/k8s/client/informers: No such file or directory
```

File: pkg/k8s/slim/README.md
Comment (tklauser) [inline pkg/k8s/slim/README.md:113]: Thanks for adding this section :heart:

**Review by tklauser — APPROVED** [review]

**Comment by christarazi** [comment]
@joestringer From my understanding, these errors are expected. I analyszed the exectuion of the script using `set -x` and traced it down to the following snippet:

```
    ( kube::codegen::internal::grep -l --null \
        -e '^// Code generated by client-gen. DO NOT EDIT.$' \
        -r "${out_dir}/${clientset_subdir}" \
        --include '*.go' \
        || true \
    ) | xargs -0 rm -f
```

It seems to me that this snippet is trying to find any files under `"${out_dir}/${clientset_subdir}"` as long as they match the regex string, then delete them, hence the `xargs rm` pipe. Notice the `true` at the end of the `grep` command, which suggests that the script should not fail if the grep fails, presumably if the path does not  exist. However, the path is not expected to exist because the subsequent step after the snippet above is to generate the code. In other words, this snippet is to remove any leftover files before beginning the code generation. These errors are expected in summary.

File: Makefile
Comment (christarazi) [reply Makefile:400]: This part was added by @aanm. Andre, is there a reason for moving the `cp` into the script?

Comment (aanm) [reply Makefile:400]: @joestringer @christarazi yes, the `cp` needs to be done before the code generation for the remaining of the packages. You can test it out by removing the `cp` from the script and re-add it in the Makefile to see the difference of the results

**Comment by aanm** [comment]
@christarazi needs rebase

File: Makefile
Comment (joestringer) [reply Makefile:400]: I see. Probably the solution to my nit is to move even more of the build process out of make, but that'd be a bigger ask. Feels messy to me, but nothing to hold up this PR over.

**Comment by christarazi** [comment]
Updated to final v1.30.0 from the rc.1. No changes in the slim structures as well.

**Review by ysksuzuki — APPROVED** [review]
CEGP CRD change looks good to me

**Pull request merged**
